spv/sync.go

```diff
@@ -1322,81 +1322,225 @@ var hashStop chainhash.Hash
 // is up to date with all connected peers.  This is part of the startup sync
 // process.
 func (s *Syncer) initialSyncHeaders(ctx context.Context) error {
-	startTime := time.Now()
+	// The strategy for fetching initial headers is to split each batch of
+	// headers into 3 stages that are run in parallel, with each batch
+	// flowing through the pipeline:
+	//
+	// 1. Fetch headers
+	// 2. Fetch cfilters
+	// 3. Chain switch
+	//
+	// In the positive path during IBD, this allows processing the chain
+	// much faster, because the bottleneck is generally the latency for
+	// completing steps 1 and 2, and thus running them in parallel for
+	// each batch speeds up the overall sync time.
+	//
+	// The pipelining is accomplished by assuming that the next batch of
+	// headers will be a successor of the most recently fetched batch and
+	// requesting this next batch using locators derived from this (not yet
+	// completely processed) batch.  For all but the most recent blocks in
+	// the main chain, this assumption will be true.
+	//
+	// Ocasionally, in case of bisheaving peers, the sync process might be
+	// entirely reset and started from scratch using only the wallet derived
+	// block locators.
+	//
+	// The three stages are run as goroutines in the following g.
+	g, ctx := errgroup.WithContext(ctx)
 
-nextbatch:
-	for ctx.Err() == nil {
-		// Fetch a batch of headers.
-		batch, err := s.getHeaders(ctx)
-		if err != nil {
-			return err
-		}
-		if batch.done {
-			// All done.
-			log.Debugf("Initial sync completed in %s",
-				time.Since(startTime).Round(time.Second))
-			return nil
-		}
+	// This chan is written to when one of the later stages (cfilter
+	// fetching or chain switch) fails and the header fetching stage should
+	// restart without using the cached chain.
+	invalidateBatchChan := make(chan struct{})
```

Comment:
[Updated](https://github.com/decred/dcrwallet/compare/a85a6b7cb82f8db272b1ad7ad32cb98b2ceeb528..8d9b54d4086a980dcdc287e017ef19b67723dbcf) to use chan closures as the signal. This is a bit more involved to ensure that the chan is not closed twice.

I have tested this by running with `-race` in mainnet and purposefully triggered the fail paths by modifying a dcrd instance to stall forever and/or return faulty cfilters.

Reply:
Creating the new channel and closing the previous should occur at the same time under the lock.

See https://go.dev/play/p/NmVKMLqKw0E

The updated code is also not rereading the channel under the mutex the next time the loop iterates, so that doesn't appear to be race safe.

Reply:
Updated. Note that I'm not storing the channel inside the loop, but only on the first iteration (and then if it closed) in order to avoid missing a close if it happens after the last channel check and before the next iteration.

Reply:
thanks, this looks correct now